Implementation of trigger front-end best-choice codec

jbsauvan opened the pull request:
Contribution to issue #7
- New features: 
  - Best choice coding scheme as developed in Split, composed of energy sums in trigger cells and selection of the 12 trigger cells with highest energies.
  - Dummy backend algorithm using this coding scheme as input and performing the sums of trigger cells in each module.
- Changes in base code:
  - Had to pass the geometry to HGCalTriggerFECodecBase::setDataPayload() in order to retrieve HGC cells from trigger cells from modules; the module class gives the list of trigger cell DetId, and the geometry is needed to get HGC cells from that (linked to issue #6 )
  - Had to change HGCFETriggerDigi::setDetId(): DetId class doesn't have a uint32_t() method.

jbsauvan committed:
First test version of the Best Choice encoding

jbsauvan committed:
Comment out debug print

jbsauvan committed:
Make HGCalBestChoiceCodec consistent with new TriggerDetId scheme

jbsauvan committed:
Add decode to BestChoice codec

jbsauvan committed:
Add config parameters for HGCalBestChoiceCodec

jbsauvan committed:
Add dummy backend algo making module sums

jbsauvan committed:
Revert back to ADC+TDC instead of ADC only

jbsauvan committed:
Cleaning a bit

lgray commented:
@jbsauvan Can you supply some testing routines (unit tests) for the new codec so that we can independently test that it functions?

lgray commented on DataFormats/L1THGCal/interface/HGCFETriggerDigi.h:
```diff
@@ -51,7 +51,7 @@ namespace l1t {
     IDTYPE getDetId() const { return IDTYPE(detid_); }
     template<typename IDTYPE>
-    void   setDetId(const IDTYPE& id) { detid_ = id.uint32_t(); }
+    void   setDetId(const IDTYPE& id) { detid_ = id(); }
```
This is wrong (it calls the codec's id() function).
All DetIds have the uint32_t() function: https://github.com/cms-sw/cmssw/blob/CMSSW_6_2_X_SLHC/DataFormats/DetId/interface/DetId.h#L42

jbsauvan replied:
Isn't it a conversion operator instead of a function?

lgray replied:
You can call it like a function too, it's just idiomatic that you can also do (uint32_t) to convert.

jbsauvan replied:
Function call doesn't compile:
`DataFormats/L1THGCal/interface/HGCFETriggerDigi.h:54:41: error: 'const class HGCTriggerDetId' has no member named 'uint32_t'`
`uint32_t(id)` and `(uint32_t)id` are working.

lgray replied:
Hmm... just use `id.rawId();` instead.

jbsauvan replied:
done

lgray commented on L1Trigger/L1THGCal/interface/HGCalTriggerFECodecBase.h:
```diff
@@ -99,7 +101,7 @@ namespace HGCalTriggerFE {
           << this->name() << " overwriting current data!";
       }
-      static_cast<Impl&>(*this).setDataPayloadImpl(mod,ee,fh,bh);
+      static_cast<Impl&>(*this).setDataPayloadImpl(mod,geom,ee,fh,bh);
```
If the module is properly defined we shouldn't need the geometry to do anything, no? All relevant DetIds all the way down to HGC cell level are available.

jbsauvan replied:
The module class gives the list of trigger cell detids (https://github.com/PFCal-dev/cmssw/blob/hgc_trigger_fwk_SLHC26p3/L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h#L99), not the TriggerCell objects. To get the HGC cells inside each trigger cell we need the geometry (or the TriggerCell objects), right?

lgray replied:
Indeed, you are correct! However, I think the correct action is to fix the geometry class itself rather than kludge things downstream of it. I'll send you a pull request in a bit.

lgray commented on L1Trigger/L1THGCal/interface/fe_codecs/HGCalBestChoiceCodec.h:
```diff
@@ -0,0 +1,61 @@
+        for(auto& value : payload)
+        {
+            value = 0;
+        }
```
This whole loop can be replaced with: payload.fill(0);

jbsauvan replied:
done

lgray commented on L1Trigger/L1THGCal/plugins/fe_codecs/HGCalBestChoiceCodec.cc:
```diff
@@ -0,0 +1,182 @@
+    {
+        if(value_id.second>nCellsInModule_) // cell number starts at 1
+        {
+            edm::LogWarning("HGCalBestChoiceCodec") 
```
This should be an exception, it represents a badly configured state of the hardware we are simulating.

jbsauvan replied:
done

jbsauvan committed:
Simplify HGCalBestChoiceDataPayload::reset()

jbsauvan committed:
Use IDTYPE::rawId() in HGCFETriggerDigi::serDetId()

lgray committed:
let modules in HGC trigger geometry know about HGC cells

jbsauvan committed:
Merge pull request #1 from lgray/trigger_fe_bestchoice

Let modules in HGC trigger geometry know about HGC cells

jbsauvan committed:
Change Warnings to Exceptions in HGCalBestChoiceCodec

jbsauvan committed:
Merge branch 'trigger_fe_bestchoice' of https://github.com/jbsauvan/cmssw into trigger_fe_bestchoice

jbsauvan commented:
I'll fix the remaining issues (remove geometry access in codec, write testing routines) next week.

lgray commented:
@jbsauvan OK -- will just leave this PR open until you're done.

jbsauvan committed:
Remove geometry access from codec interface

jbsauvan committed:
Tests for BestChoice.

jbsauvan committed:
Add BestChoice test config

jbsauvan committed:
Move codec implementations in src/fe_codecs/

jbsauvan committed:
Add histos for HGC cells

jbsauvan commented:
@lgray I have moved the fe coded implementations in src/fe_codecs, and modified the tests accordingly. Tests are now compiling and running, there is one unit test checking the consistency of the encoding and the decoding, and one test plugin that produces histograms like the number of cells per modules, the cell data values, the sums of data values in modules, for output trigger cells and input HGC cells.

lgray commented:
@jbsauvan awesome!

lgray commented:
Indeed a quick pass of the code look good. I'll take a closer look soon.

lgray commented:
@jbsauvan Now taking a look, comments soon.

lgray commented on L1Trigger/L1THGCal/test/HGCalTriggerBestChoiceTester.cc:
```diff
@@ -0,0 +1,191 @@
+        std::unique_ptr<HGCalBestChoiceCodecImpl> codec_;
+        edm::Service<TFileService> fs_;
+        // histos
+        TH1F* hgcCellsPerModule_;
```
make these `std::unique_ptr<TH1F>`, right now you are leaking memory.

lgray replied:
Ah, never mind. TFileService.

lgray commented:
@jbsauvan Looks good, all tests run and produce some reasonable output.
Merging.
